File: sections/98_plug_treewidth_frontiers/problems/b_connected_cell_sets/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
const int MOD=1e9+7;
using State=pair<vector<int>,bool>;
vector<int>canon(vector<int>a) {
    map<int,int>mp;
    int nxt=1;
    for(int&x:a)if(x) {
        if(!mp.count(x))mp[x]=nxt++;
        x=mp[x];
    }
    return a;
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int h,w;
    cin>>h>>w;
    vector<string>g(h);
    for(auto&s:g)cin>>s;
    map<State,int>dp;
    dp[ {
        vector<int>(w),false
    }
    ]=1;
    for(int r=0;r<h;r++)for(int c=0;c<w;c++) {
        map<State,int>nd;
        for(auto[state,value]:dp) {
            auto labels=state.first;
            bool closed=state.second;
            int up=labels[c],left=c?labels[c-1]:0;
            auto z=labels;
            z[c]=0;
            bool vanished=up&&find(z.begin(),z.end(),up)==z.end();
            if(!vanished||(!closed&&count_if(z.begin(),z.end(),[](int x){return x;})==0)) {
                State key= {
                    canon(z),closed||vanished
                }
                ;
                nd[key]=(nd[key]+value)%MOD;
            }
            if(g[r][c]=='.'&&!closed) {
                z=labels;
                if(left&&up&&left!=up)for(int&x:z)if(x==up)x=left;
                z[c]=left?left:(up?up:*max_element(z.begin(),z.end())+1);
                State key= {
                    canon(z),false
                }
                ;
                nd[key]=(nd[key]+value)%MOD;
            }
        }
        dp.swap(nd);
    }
    ll answer=0;
    for(auto[state,value]:dp) {
        set<int>s(state.first.begin(),state.first.end());
        s.erase(0);
        if(state.second||s.size()==1)answer+=value;
    }
    cout<<answer%MOD<<'\n';
}

File: sections/98_plug_treewidth_frontiers/problems/b_connected_cell_sets/solution.py
import sys
MOD = 1000000007

def canon(a):
    mp = {}
    nxt = 1
    out = []
    for x in a:
        if x and x not in mp:
            mp[x] = nxt
            nxt += 1
        out.append(mp.get(x, 0))
    return tuple(out)

def main():
    data = sys.stdin.buffer.read().split()
    h, w = map(int, data[:2])
    g = [x.decode() for x in data[2:]]
    dp = {(tuple([0] * w), False): 1}
    for r in range(h):
        for c in range(w):
            nd = {}
            for (labels, closed), value in dp.items():
                old = list(labels)
                up = old[c]
                left = old[c - 1] if c else 0
                new = old[:]
                new[c] = 0
                vanished = up and up not in new
                if not vanished or (not any(new) and (not closed)):
                    key = (canon(new), closed or bool(vanished))
                    nd[key] = (nd.get(key, 0) + value) % MOD
                if g[r][c] == '.' and (not closed):
                    new = old[:]
                    if left and up and (left != up):
                        new = [left if x == up else x for x in new]
                    new[c] = left or up or max(new, default=0) + 1
                    key = (canon(new), False)
                    nd[key] = (nd.get(key, 0) + value) % MOD
            dp = nd
    answer = 0
    for (labels, closed), value in dp.items():
        if closed or len(set(labels) - {0}) == 1:
            answer = (answer + value) % MOD
    print(answer)
if __name__ == '__main__':
    main()
